composer/datasets/streaming/writer.py

```diff
@@ -60,9 +61,14 @@ class StreamingDatasetWriter(object):
         dirname (str): Directory to write shards to.
         fields: (List[str]): The fields to save for each sample.
         shard_size_limit (int): Maximum shard size in bytes. Default: `1 << 24`.
+        compression (Optional[str]): Compression algorithm and optional compression level. Currently supported: 'gz', 'gz:[1-9]' or None.
     """
 
-    def __init__(self, dirname: str, fields: List[str], shard_size_limit: int = 1 << 24) -> None:
+    def __init__(self,
+                 dirname: str,
+                 fields: List[str],
+                 shard_size_limit: int = 1 << 24,
+                 compression: Optional[str] = 'gz') -> None:
```

Comment:
CC @abhi-mosaic, but should this default to gz or None? I think I would prefer None because of the additional compression/decompression overhead it can add, but can be convinced otherwise if using compression is strictly better. (i.e. we would still use compression on clouds where we would have zero egress and storage fees)

Reply:
My intuition here is that on an ML box it would take longer to {up/down}load even a medium sized dataset than to [de]compress it -- happy to run benchmarks after this lands to test things though.